File: GDTMobApp/Release/GDTAds/GDTUnifiedNativeAd/ViewController/UnifiedNativeAdPortraitDetailViewController.m
//
//  UnifiedNativeAdPortraitDetailViewController.m
//  GDTMobApp
//
//  Created by royqpwang on 2019/6/2.
//  Copyright © 2019 Tencent. All rights reserved.
//

#import "UnifiedNativeAdPortraitDetailViewController.h"
#import "UnifiedNativeAdCustomView.h"
#import "DemoUtil.h"

@interface UnifiedNativeAdPortraitDetailViewController () <GDTUnifiedNativeAdViewDelegate>

@property (nonatomic, strong) UnifiedNativeAdCustomView *nativeAdView;
@property (nonatomic, strong) NSLayoutConstraint *consH;

@end

@implementation UnifiedNativeAdPortraitDetailViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.view.backgroundColor = [UIColor blackColor];
    [self initViews];
    self.nativeAdView.viewController = self;
    self.nativeAdView.delegate = self;
    
    if ([self.dataObject isAdValid]) {
        [self.nativeAdView registerDataObject:self.dataObject clickableViews:@[self.nativeAdView.imageView,
                                                                               self.nativeAdView.titleLabel,
                                                                               self.nativeAdView.iconImageView,
                                                                               self.nativeAdView.descLabel,
                                                                               self.nativeAdView.clickButton]];
    }
    
    [self.nativeAdView setupWithUnifiedNativeAdObject:self.dataObject];
    
    if (self.nativeAdView.dataObject.imageWidth > 0) {
        self.consH.active = NO;
        self.consH = [self.nativeAdView.mediaView.heightAnchor constraintEqualToAnchor:self.nativeAdView.mediaView.widthAnchor multiplier:(float)self.nativeAdView.dataObject.imageHeight/(float)self.nativeAdView.dataObject.imageWidth];
        self.consH.active = YES;
    }
    
    [self.dataObject bindImageViews:@[self.nativeAdView.imageView] placeholder:nil];
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    [self.nativeAdView.mediaView play];
}

- (void)initViews
{
    [self.view addSubview:self.nativeAdView];
    self.nativeAdView.translatesAutoresizingMaskIntoConstraints = NO;
    [self.nativeAdView.leftAnchor constraintEqualToAnchor:self.view.leftAnchor].active = YES;
    [self.nativeAdView.rightAnchor constraintEqualToAnchor:self.view.rightAnchor].active = YES;
    [self.nativeAdView.topAnchor constraintEqualToAnchor:self.view.topAnchor].active = YES;
    [self.nativeAdView.bottomAnchor constraintEqualToAnchor:self.view.bottomAnchor].active = YES;
    // mediaView
    self.nativeAdView.imageView.translatesAutoresizingMaskIntoConstraints = NO;
    [self.nativeAdView.imageView.widthAnchor constraintEqualToAnchor:self.view.widthAnchor].active = YES;
    [self.nativeAdView.imageView.heightAnchor constraintEqualToAnchor:self.view.heightAnchor].active = YES;
    [self.nativeAdView.imageView.centerXAnchor constraintEqualToAnchor:self.view.centerXAnchor].active = YES;
    [self.nativeAdView.imageView.centerYAnchor constraintEqualToAnchor:self.view.centerYAnchor].active = YES;
    self.nativeAdView.imageView.alpha = 0.4;
    
    self.nativeAdView.clickButton.translatesAutoresizingMaskIntoConstraints = NO;
    [self.nativeAdView.clickButton.leftAnchor constraintEqualToAnchor:self.nativeAdView.leftAnchor constant:15].active = YES;
    [self.nativeAdView.clickButton.rightAnchor constraintEqualToAnchor:self.nativeAdView.rightAnchor constant:-20].active = YES;
    [self.nativeAdView.clickButton.bottomAnchor constraintEqualToAnchor:self.nativeAdView.bottomAnchor constant:-15].active = YES;
    [self.nativeAdView.clickButton.heightAnchor constraintEqualToConstant:44].active = YES;
    self.nativeAdView.clickButton.backgroundColor = [UIColor orangeColor];
    self.nativeAdView.descLabel.translatesAutoresizingMaskIntoConstraints = NO;
    self.nativeAdView.descLabel.textColor = [UIColor redColor];
    [self.nativeAdView.descLabel.leftAnchor constraintEqualToAnchor:self.nativeAdView.clickButton.leftAnchor].active = YES;
    [self.nativeAdView.descLabel.rightAnchor constraintEqualToAnchor:self.nativeAdView.clickButton.rightAnchor].active = YES;
    [self.nativeAdView.descLabel.heightAnchor constraintEqualToConstant:30].active = YES;
    [self.nativeAdView.descLabel.bottomAnchor constraintEqualToAnchor:self.nativeAdView.clickButton.topAnchor constant:-10].active = YES;
    self.nativeAdView.titleLabel.translatesAutoresizingMaskIntoConstraints = NO;
    self.nativeAdView.titleLabel.textColor = [UIColor blueColor];
    [self.nativeAdView.titleLabel.leftAnchor constraintEqualToAnchor:self.nativeAdView.clickButton.leftAnchor].active = YES;
    [self.nativeAdView.titleLabel.rightAnchor constraintEqualToAnchor:self.nativeAdView.clickButton.rightAnchor].active = YES;
    [self.nativeAdView.titleLabel.heightAnchor constraintEqualToConstant:30].active = YES;
    [self.nativeAdView.titleLabel.bottomAnchor constraintEqualToAnchor:self.nativeAdView.descLabel.topAnchor constant:-10].active = YES;
    self.nativeAdView.iconImageView.translatesAutoresizingMaskIntoConstraints = NO;
    [self.nativeAdView.iconImageView.leftAnchor constraintEqualToAnchor:self.nativeAdView.clickButton.leftAnchor].active = YES;
    [self.nativeAdView.iconImageView.heightAnchor constraintEqualToConstant:60].active = YES;
    [self.nativeAdView.iconImageView.widthAnchor constraintEqualToConstant:60].active = YES;
    [self.nativeAdView.iconImageView.bottomAnchor constraintEqualToAnchor:self.nativeAdView.titleLabel.topAnchor constant:-10].active = YES;
    [self.nativeAdView addSubview:self.nativeAdView.logoView];
    self.nativeAdView.logoView.translatesAutoresizingMaskIntoConstraints = NO;
    [self.nativeAdView.logoView.widthAnchor constraintEqualToConstant:kGDTLogoImageViewDefaultWidth].active = YES;
    [self.nativeAdView.logoView.heightAnchor constraintEqualToConstant:kGDTLogoImageViewDefaultHeight].active = YES;
    [self.nativeAdView.logoView.rightAnchor constraintEqualToAnchor:self.nativeAdView.rightAnchor].active = YES;
    [self.nativeAdView.logoView.bottomAnchor constraintEqualToAnchor:self.nativeAdView.bottomAnchor constant:-20].active = YES;
    // mediaView
    self.nativeAdView.mediaView.translatesAutoresizingMaskIntoConstraints = NO;
    [self.nativeAdView.mediaView.widthAnchor constraintEqualToAnchor:self.view.widthAnchor].active = YES;
    self.consH = [self.nativeAdView.mediaView.heightAnchor constraintEqualToAnchor:self.view.heightAnchor];
    self.consH.active = YES;
    [self.nativeAdView.mediaView.centerXAnchor constraintEqualToAnchor:self.view.centerXAnchor].active = YES;
    [self.nativeAdView.mediaView.centerYAnchor constraintEqualToAnchor:self.view.centerYAnchor constant:-60].active = YES;
}

#pragma mark - GDTUnifiedNativeAdViewDelegate
- (void)gdt_unifiedNativeAdViewWillExpose:(GDTUnifiedNativeAdView *)unifiedNativeAdView
{
    NSLog(@"%s",__FUNCTION__);
    NSLog(@"%@ 广告曝光", unifiedNativeAdView.dataObject.title);
}

- (void)gdt_unifiedNativeAdDetailViewWillPresentScreen:(GDTUnifiedNativeAdView *)unifiedNativeAdView
{
    NSLog(@"%s",__FUNCTION__);
    [self.nativeAdView.mediaView pause];
}

- (void)gdt_unifiedNativeAdDetailViewClosed:(GDTUnifiedNativeAdView *)unifiedNativeAdView
{
    NSLog(@"%s",__FUNCTION__);
    [self.nativeAdView.mediaView play];
}

- (void)gdt_unifiedNativeAdView:(GDTUnifiedNativeAdView *)unifiedNativeAdView playerStatusChanged:(GDTMediaPlayerStatus)status userInfo:(NSDictionary *)userInfo
{
    NSString *statusString = [DemoUtil videoPlayerStatusStringFromStatus:status];
    NSLog(@"%s-----status:%@",__FUNCTION__,statusString);
}

- (void)gdtAdComplainSuccess:(id)ad {
    NSLog(@"%s",__FUNCTION__);
    NSLog(@"广告投诉成功");
}

#pragma mark - property getter
- (UnifiedNativeAdCustomView *)nativeAdView
{
    if (!_nativeAdView) {
        _nativeAdView = [[UnifiedNativeAdCustomView alloc] init];
        _nativeAdView.accessibilityIdentifier = @"nativeAdView_id";
    }
    return _nativeAdView;
}

@end

File: GDTMobSample-Swift/GDTAdTypes/UnifiedNativeAdViewController/ViewController/UnifiedNativeAdPortraitDetailViewController.swift
//
//  UnifiedNativeAdPortraitDetailViewController.swift
//  GDTMobSample-Swift
//
//  Created by 胡城阳 on 2019/10/29.
//  Copyright © 2019 Tencent. All rights reserved.
//

import UIKit

class UnifiedNativeAdPortraitDetailViewController:UIViewController,GDTUnifiedNativeAdViewDelegate{
    
    var dataObject:GDTUnifiedNativeAdDataObject? = nil
    let nativeAdView:UnifiedNativeAdCustomView = {
        let _nativeAdView = UnifiedNativeAdCustomView()
        return _nativeAdView
    }()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        initViews()
        self.nativeAdView.viewController = self
        self.nativeAdView.delegate = self
        self.nativeAdView.registerDataObject(self.dataObject, clickableViews: [self.nativeAdView.titleLabel,self.nativeAdView.iconImageView,self.nativeAdView.descLabel,self.nativeAdView.clickButton])
        self.nativeAdView.setupWithUnifiedNativeAdObject(unifiedNativeDataObject: dataObject!)
        // Do any additional setup after loading the view.
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        self.nativeAdView.mediaView.play()
    }
    
    func initViews(){
        self.view.addSubview(nativeAdView)
        self.nativeAdView.translatesAutoresizingMaskIntoConstraints = false
        self.nativeAdView.leftAnchor.constraint(equalTo: self.view.leftAnchor).isActive = true
        self.nativeAdView.rightAnchor.constraint(equalTo: self.view.rightAnchor).isActive = true
        self.nativeAdView.topAnchor.constraint(equalTo: self.view.topAnchor).isActive = true
        self.nativeAdView.bottomAnchor.constraint(equalTo: self.view.bottomAnchor).isActive = true
        
        self.nativeAdView.clickButton.translatesAutoresizingMaskIntoConstraints = false
        self.nativeAdView.clickButton.leftAnchor.constraint(equalTo: self.nativeAdView.leftAnchor, constant: 15).isActive = true
        self.nativeAdView.clickButton.rightAnchor.constraint(equalTo: self.nativeAdView.rightAnchor, constant: -20).isActive = true
        self.nativeAdView.clickButton.bottomAnchor.constraint(equalTo: self.nativeAdView.bottomAnchor, constant: -15).isActive = true
        self.nativeAdView.clickButton.heightAnchor.constraint(equalToConstant: 44).isActive = true
        self.nativeAdView.clickButton.backgroundColor = UIColor.orange
        
        self.nativeAdView.descLabel.translatesAutoresizingMaskIntoConstraints = false
        self.nativeAdView.descLabel.textColor = UIColor.red
        self.nativeAdView.descLabel.leftAnchor.constraint(equalTo: self.nativeAdView.clickButton.leftAnchor).isActive = true
        self.nativeAdView.descLabel.rightAnchor.constraint(equalTo: self.nativeAdView.clickButton.rightAnchor).isActive = true
        self.nativeAdView.descLabel.heightAnchor.constraint(equalToConstant: 30).isActive = true
        self.nativeAdView.descLabel.bottomAnchor.constraint(equalTo: self.nativeAdView.clickButton.topAnchor, constant: -10).isActive = true

        self.nativeAdView.titleLabel.translatesAutoresizingMaskIntoConstraints = false
        self.nativeAdView.titleLabel.textColor = UIColor.blue
        self.nativeAdView.titleLabel.leftAnchor.constraint(equalTo: self.nativeAdView.clickButton.leftAnchor).isActive = true
        self.nativeAdView.titleLabel.rightAnchor.constraint(equalTo: self.nativeAdView.clickButton.rightAnchor).isActive = true
        self.nativeAdView.titleLabel.heightAnchor.constraint(equalToConstant: 30).isActive = true
        self.nativeAdView.titleLabel.bottomAnchor.constraint(equalTo: self.nativeAdView.descLabel.topAnchor, constant: -10).isActive = true
        
        self.nativeAdView.iconImageView.translatesAutoresizingMaskIntoConstraints = false
        self.nativeAdView.iconImageView.leftAnchor.constraint(equalTo: self.nativeAdView.clickButton.leftAnchor).isActive = true
        self.nativeAdView.iconImageView.heightAnchor.constraint(equalToConstant: 60).isActive = true
        self.nativeAdView.iconImageView.widthAnchor.constraint(equalToConstant: 60).isActive = true
    self.nativeAdView.iconImageView.bottomAnchor.constraintEqualToSystemSpacingBelow(self.nativeAdView.titleLabel.topAnchor, multiplier: -10).isActive = true
        
        self.nativeAdView.addSubview(self.nativeAdView.logoView)
        self.nativeAdView.logoView.translatesAutoresizingMaskIntoConstraints = false
        self.nativeAdView.logoView.widthAnchor.constraint(equalToConstant: kGDTLogoImageViewDefaultWidth).isActive = true
        self.nativeAdView.logoView.heightAnchor.constraint(equalToConstant: kGDTLogoImageViewDefaultHeight).isActive = true
        self.nativeAdView.logoView.rightAnchor.constraint(equalTo: self.nativeAdView.rightAnchor).isActive = true
        self.nativeAdView.logoView.bottomAnchor.constraintEqualToSystemSpacingBelow(self.nativeAdView.bottomAnchor, multiplier: -20).isActive = true
    }
    
}


extension UnifiedNativeAdPortraitDetailViewController{
    func gdt_unifiedNativeAdViewWillExpose(_ unifiedNativeAdView: GDTUnifiedNativeAdView!) {
        print(#function)
        print("\(unifiedNativeAdView.dataObject.title ?? "") 广告曝光")
    }
    
    func gdt_unifiedNativeAdDetailViewWillPresentScreen(_ unifiedNativeAdView: GDTUnifiedNativeAdView!) {
        print(#function)
        self.nativeAdView.mediaView.pause()
    }
    
    func gdt_unifiedNativeAdDetailViewClosed(_ unifiedNativeAdView: GDTUnifiedNativeAdView!) {
        print(#function)
        self.nativeAdView.mediaView.play()
    }
    
    func gdt_unifiedNativeAdView(_ unifiedNativeAdView: GDTUnifiedNativeAdView!, playerStatusChanged status: GDTMediaPlayerStatus, userInfo: [AnyHashable : Any]! = [:]) {
        print(#function)
    }
}
